### DISCUSSION TITLE
Infinispan error during deployments and sessions lost

### DISCUSSION THREAD

**Discussion by anand-kulk** [keycloak/keycloak#13909] [Q&A]
Hello All-
   I am using Keycloak 16 bitnami chart in HA deployment in kubernetes with 5 replicas(pods) each running on separate node.
I am using default infinispan(running on keycloak pods) cache with  4 owners for cache. Using KUBE_ping for cluster sync. 

`cache:
  authOwnersCount: 4
  ownersCount: 4`

I had a maintenance to do for 2 nodes and I drained them and evicted 2 of these 5 Keycloak pods. I was hoping this to not have any downtime as in theory I still had 2 more pods of the remaining that would have replicated cache. But I saw below errors and also lost all user sessions. Admin panel was down and wasn't able to get in, I ended up restarting the whole cluster to resolve this.

_org.infinispan.util.concurrent.TimeoutException: ISPN000476: Timed out waiting for responses for request 1592914 from keycloak-4,keycloak-2_ 

Given my goal is to have zero downtime and no session loss, could someone answer my concerns below -

1. If pods go down, isn't there automatic re-balancing in distributed caching that it won't try to reach dead pods? 
2. Also, not sure why 4 cache owners were not enough to prevent downtime.
3. Researching more, it sounds like using separate cluster for inifinispan or offline sessions is a probable solution. Would upgrading to Keycloak 19(quarkus based) resolve any of these issues so I don't need to do separate infinispan cluster setup?

**Comment by ddurham2** [comment]
@anand-kulk, did you ever figure things out?  The loss of user-sessions _really_ concerns me about keycloak (whether it be upon upgrade or just because the system is under load).  Thanks

**Reply by anand-kulk** [reply]
Hi @ddurham2 - apologies for getting back late. In case it helps, this is what I did to avoid loss of sessions-
Upgraded to Keycloak 19 and used remote infinispan caching cluster that used JDBC store for storing cache entries. Since that was done, I haven't seen any issues. Architecturally, it is better to have caching logic reside outside Keycloak process. Let me know if any questions.

**Reply by manologg** [reply]
Is there a particular reason why you updated to Keycloak 19? I'm currently on Keycloak 18 and facing the same issues

**Reply by manologg** [reply]
Also: where can you find a good guide on how to use a remote infinispan cluster? Or can you share your configuration? That would be really helpful :smiley:

**Reply by anand-kulk** [reply]
@manologg - I believe that was the latest Quarkus version available when I did upgrade. I don't think it is much related to Keycloak version but about how you handle your caching.

**Reply by anand-kulk** [reply]
@manologg - Are you in Kubernetes? Using infinispan cluster is 2 step process, first deploy infinispan cluster. I initially used Infinispan operator, but I won't recommend that, you can use Helm chart https://github.com/infinispan/infinispan-helm-charts.
You can refer https://infinispan.org/documentation/ on how to configure caches. I am using remote cache for user sessions and client sessions.
Second step, once infinispan cluster is up, keycloak needs to use remote ispn configuration and connect to infinispan hot rod service, ex configuration-
`<distributed-cache name="sessions" owners="${cache-owners-count}">
            <expiration lifespan="-1"/>
            <remote-store cache="sessions" xmlns="urn:infinispan:config:store:remote:13.0"
                          fetch-state="false"
                          purge="false"
                          preload="false"
                          shared="true" segmented="false"
                          connect-timeout="${ispn-connect-timeout:2000}">
                <remote-server host="${ispn-hostname:infinispan}" port="${infinispan.bind.port}"/>
                <security>
                    <authentication>
                        <digest username="${ispn-username}"
                                password="${ispn-password}"
                                realm="default"/>
                    </authentication>
                </security>
                <property name="rawValues">true</property>
                <property name="marshaller">org.keycloak.cluster.infinispan.KeycloakHotRodMarshallerFactory</property>
            </remote-store>
        </distributed-cache>
        <distributed-cache name="clientSessions" owners="${cache-owners-count}">
            <expiration lifespan="-1"/>
            <remote-store cache="clientSessions" xmlns="urn:infinispan:config:store:remote:13.0"
                          fetch-state="false"
                          purge="false"
                          preload="false"
                          shared="true" segmented="false"
                          connect-timeout="${ispn-connect-timeout:2000}">
                <remote-server host="${ispn-hostname}" port="${infinispan.bind.port}"/>
                <security>
                    <authentication>
                        <digest username="${ispn-username}"
                                password="${ispn-password}"
                                realm="default"/>
                    </authentication>
                </security>
                <property name="rawValues">true</property>
                <property name="marshaller">org.keycloak.cluster.infinispan.KeycloakHotRodMarshallerFactory</property>
            </remote-store>
        </distributed-cache>`

**Comment by dglozano** [comment]
We are experiencing the same issue. 4 owners and doing a rolling deploy of one node at time nearly always results in a complete lost of sessions. We have 4 tasks running in AWS ECS Fargate and configured with JDBC_PING. We are now considering setting up a remote infinispan caching cluster as anand-kulk mentioned above.
